nf_core/components/list.py

```diff
@@ -0,0 +1,146 @@
+import json
+import logging
+
+import rich
+
+import nf_core.modules.modules_utils
+from nf_core.components.components_command import ComponentCommand
+from nf_core.modules.modules_json import ModulesJson
+
+# from .modules_command import ModulesRepo
+from nf_core.modules.modules_repo import ModulesRepo
+
+log = logging.getLogger(__name__)
+
+
+class ComponentList(ComponentCommand):
+    def __init__(self, component_type, pipeline_dir, remote=True, remote_url=None, branch=None, no_pull=False):
+        super().__init__(component_type, pipeline_dir, remote_url, branch, no_pull)
+        self.remote = remote
+
+    def list_components(self, keywords=None, print_json=False):
+        """
+        Get available modules/subworkflows names from GitHub tree for repo
+        and print as list to stdout
+        """
+        # Check modules directory structure
+        # self.check_component_structure(self.component_type)
+
+        # Initialise rich table
+        table = rich.table.Table()
+        table.add_column(f"{self.component_type[:-1].capitalize()} Name")
+        components = []
+
+        if keywords is None:
+            keywords = []
+
+        def pattern_msg(keywords):
+            if len(keywords) == 0:
+                return ""
+            if len(keywords) == 1:
+                return f" matching pattern '{keywords[0]}'"
+            else:
+                quoted_keywords = (f"'{key}'" for key in keywords)
+                return f" matching patterns {', '.join(quoted_keywords)}"
+
```

Comment:
I'm never sure on when it is really necessary to have a method within a method, but this feels like it could (and should) be a standalone method outside the class. I find it particularly confusing that `keywords` is used in the scope of this internal method whilst being defined in the enclosing method directly above.